### PR TITLE
[Bug Report] Concurrency problems and low performance when obtaining financial reports

### LINKED ISSUE #3759: [Bug Report] Concurrency problems and low performance when obtaining financial reports
## Bug report

The current design for obtaining financial reports does not correctly handle transactions, since if a report requires minutes to obtain it, the system locks the T_Report table and queues the report requests until the first one that was executed finishes.

On the other hand, it is intended to review how records are built and inserted in the temporary table since in some reports it takes too long.

#### Steps to reproduce
Open several sessions on different computers and send a statement of results, after a year of operation the process simply does not end.

#### Expected behavior
Multiple users should be able to send financial reports without blocking issues and poor performance in retrieving information.

### Approach

- Review the current approach based on a temporary table
- Improve transaction handling by not using the same process transaction
- Perform a query optimization to get the reports, DBA work.  

#### Other relevant information
- Your OS: MacOS
- Java version: 11
- ADempiere version: 394 based on develop
- Browser Version [e.g. 22]
- DB Provider: PostgreSQL y Oracle
- DB version: PG 14.2 and Oracle 18

### PR DESCRIPTION
**Pull request by e-Evolution**
fixed #3759

### CODE REVIEW

**Review by marcalwestf — CHANGES_REQUESTED** [review]
Test with Financial Report
- Name: _Income Statement Current Month_
- Lines: _Income Statement_
- Columns: only one from _Budget YTD_
- Display Line Sources: **YES**

Result:
imagen

Reason:
Method org.compiere.report.FinReport.insertLineSource() creates the SQL wrong:
imagen

After the value of column _LevelNo_, a comma character (**,**)  is missing before the calculation of _Col_0_'s value.

Adding the comma to the SQL, and running it on PGAdmin, yields no error.

When the field _**List Sources**_ is set to **_FALSE_**, then the report is generated correctly:
imagen

**Review by SusanneCalderon — CHANGES_REQUESTED** [review]
It works only without line sources. 
When line sources are included, either in the definition of the report source, or directly, when creating the report, this error occurs.
imagen

Reason: Class FinReport, line 1468
```

     StringBuilder insertLineSource = new StringBuilder("INSERT INTO T_Report "
                    + "(AD_PInstance_ID, PA_ReportLine_ID, Record_ID,Fact_Acct_ID,LevelNo ");

```      
After column _LevelNo,_ no comma is included.

**Comment by SusanneCalderon** [comment]
@e-Evolution,  it would increase the efficiency or reviewing when you separate name refactoring from code adjustment. Thanks

**Review by rcapecce — CHANGES_REQUESTED** [review]
Test:

Financial Report: Balance Sheet Current Month
Org: HQ
Definition:
unknown2



Result: We found the same message that @marcalwestf and @SusanneCalderon running the financial report

unknown

File: base/src/org/compiere/report/FinReport.java [hunk 1]
```diff
@@ -951,325 +1383,531 @@ && getListSources().equals(MReport.LISTSOURCES_No)) {
...
+                StringBuilder groupBy = new StringBuilder("");
+                for (int j = 0; j < colNames.size(); j++) {
+                    groupBy.append(", ").append(colNames.get(j));
+                }
+                insertLineSource.append(groupBy.toString().replaceFirst(", ", ""));
+            } else {
+                insertLineSource.append(variable);
+            }
+            //
+            int no = DB.executeUpdateEx(insertLineSource.toString(), trxName);
```
Comment (rcapecce) [inline base/src/org/compiere/report/FinReport.java:1570]: The stacktrace shows in this insert a sql error

```
org.postgresql.util.PSQLException: ERROR: error de sintaxis en o cerca de ?(?
  Position: 254; State=42601; ErrorCode=0
        at org.postgresql.core.v3.QueryExecutorImpl.receiveErrorResponse(QueryExecutorImpl.java:2440)
        at org.postgresql.core.v3.QueryExecutorImpl.processResults(QueryExecutorImpl.java:2183)
        at org.postgresql.core.v3.QueryExecutorImpl.execute(QueryExecutorImpl.java:308)
        at org.postgresql.jdbc.PgStatement.executeInternal(PgStatement.java:441)
        at org.postgresql.jdbc.PgStatement.execute(PgStatement.java:365)
        at org.postgresql.jdbc.PgPreparedStatement.executeWithFlags(PgPreparedStatement.java:143)
        at org.postgresql.jdbc.PgPreparedStatement.executeUpdate(PgPreparedStatement.java:120)
        at com.mchange.v2.c3p0.impl.NewProxyPreparedStatement.executeUpdate(NewProxyPreparedStatement.java:105)
        at sun.reflect.GeneratedMethodAccessor47.invoke(Unknown Source)
        at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
        at java.lang.reflect.Method.invoke(Method.java:497)
        at org.compiere.db.StatementProxy.invoke(StatementProxy.java:100)
        at com.sun.proxy.$Proxy4.executeUpdate(Unknown Source)
        at org.compiere.util.DB.executeUpdateEx(DB.java:1098)
        at org.compiere.util.DB.executeUpdateEx(DB.java:1184)
        at org.compiere.util.DB.executeUpdateEx(DB.java:1175)
        at org.compiere.report.FinReport.lambda$insertLineSource$20(FinReport.java:1568)
        at org.compiere.report.FinReport$$Lambda$65/1681504960.run(Unknown Source)
        at org.compiere.util.Trx.run(Trx.java:527)
        at org.compiere.util.Trx.run(Trx.java:495)
        at org.compiere.report.FinReport.insertLineSource(FinReport.java:1465)
        at org.compiere.report.FinReport.insertLineSource(FinReport.java:1289)
        at org.compiere.report.FinReport.lambda$insertLineDetail$16(FinReport.java:1168)
        at org.compiere.report.FinReport$$Lambda$64/656323788.run(Unknown Source)
        at org.compiere.util.Trx.run(Trx.java:527)
        at org.compiere.util.Trx.run(Trx.java:495)
        at org.compiere.report.FinReport.insertLineDetail(FinReport.java:1164)
        at org.compiere.report.FinReport.doIt(FinReport.java:151)
        at org.compiere.process.SvrProcess.process(SvrProcess.java:175)
        at org.compiere.process.SvrProcess.startProcess(SvrProcess.java:128)

```

**Review by rcapecce — CHANGES_REQUESTED** [review]
Test:

Financial Report: Balance Sheet Current Month
Org: HQ
Definition:
unknown2



Result: We found the same message that @marcalwestf and @SusanneCalderon running the financial report

unknown

The stacktrace

**Comment by e-Evolution** [comment]
Hi @marcalwestf Mario, @SusanneCalderon Susanne, @niclopup Nico, @rcapecce Raul, can please test again, I fix the problem with the functionality valid combination, when was integrated from ADAXA contribution not was testing intensively.

**Review by marcalwestf — APPROVED** [review]
- Run the standard Garden World financial reports Balance _Sheet Current Month_, _Income Statement Current Month_ and _Statement of Cash Flows_
- The reports were defined with different values for the following fields: _Report Column Set_, _Cube_, _List Sources_, _List Transactions_
- The reports were run with different values for the following parameters: _Period_, _Cube_, _Details/Source First_
- All reports run without error and displayed a report
- The reports are consistent with the underlying data

Test passed.

**Still, for this pull request to be merged, the tests of data accuracy by** @SusanneCalderon  and @rcapecce **have to be executed**.

**Review by SusanneCalderon — APPROVED** [review]
Tested with a financial report with detail lines successfully. 
imagen

**Pull request merged**